### PR TITLE
Switch: Add support for physical keyboard and mouse (Ready to merge)

### PR DESCRIPTION
**Pull request by rsn8887**
Tested working now and ready to merge!

### CODE REVIEW

**Comment by rsn8887** [comment]
Is there a libnx function for checking whether physical keyboard or mouse are connected? It would help.

**Comment by rsn8887** [comment]
This can be merged. It works just fine now.

File: src/video/switch/SDL_switchkeyboard.c [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+#include <switch.h>
+
+#include "SDL_events.h"
+#include "SDL_log.h"
+#include "SDL_switchvideo.h"
+#include "SDL_switchkeyboard.h"
+#include "../../events/SDL_keyboard_c.h"
+
+Uint8 locks = 0;
+bool keystate[160] = { 0 };
```
Comment (fincs) [inline src/video/switch/SDL_switchkeyboard.c]: This `160` constant is hardcoded everywhere. Can it be refactored to use e.g. a `#define`?

File: src/video/switch/SDL_switchkeyboard.c [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+#include <switch.h>
+
+#include "SDL_events.h"
+#include "SDL_log.h"
+#include "SDL_switchvideo.h"
+#include "SDL_switchkeyboard.h"
+#include "../../events/SDL_keyboard_c.h"
+
+Uint8 locks = 0;
```
Comment (fincs) [inline src/video/switch/SDL_switchkeyboard.c]: `u8` or `uint8_t`.

File: src/video/switch/SDL_switchkeyboard.c [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+{
+	// We skip polling keyboard if no window is created
+	if (SDL_GetFocusWindow() == NULL)
+		return;
+
+	for (int i = 0; i < 160; i++) {
+
+		int keyCode = switch_scancodes[i];
+
+		if ((hidKeyboardDown(keyCode) || hidKeyboardHeld(keyCode)) && !keystate[i]) {
```
Comment (fincs) [inline src/video/switch/SDL_switchkeyboard.c]: If down is true then held is also true. Simplify this to `hidKeyboardHeld(keyCode) && !keystate[i]`

File: src/video/switch/SDL_switchkeyboard.c [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+					} else {
+						SDL_SendKeyboardKey(SDL_RELEASED, keyCode);
+						locks &= ~0x4;
+					}
+					break;
+				default:
+					SDL_SendKeyboardKey(SDL_PRESSED, keyCode);
+			}
+			keystate[i] = true;
+		} else if ((hidKeyboardUp(keyCode) || !hidKeyboardHeld(keyCode)) && keystate[i]) {
```
Comment (fincs) [inline src/video/switch/SDL_switchkeyboard.c]: If up is true then held is false. Simplify this to `!hidKeyboardHeld(keyCode) && keystate[i]`

File: src/video/switch/SDL_switchmouse.c [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	Uint64 buttons;
+	Uint64 changed_buttons;
+	MousePosition mouse_pos;
+	int x,y;
+
+	// We skip polling mouse if no window is created
+	if (window == NULL)
+		return;
+
+	buttons = hidMouseButtonsHeld() | hidMouseButtonsDown();
```
Comment (fincs) [inline src/video/switch/SDL_switchmouse.c]: Same thing here. Simplify this to `buttons = hidMouseButtonsHeld()`

**Review by fincs — CHANGES_REQUESTED** [review]
.

File: src/video/switch/SDL_switchkeyboard.c (hunk 3 above)
Comment (rsn8887) [reply src/video/switch/SDL_switchkeyboard.c]: Thanks for the review. This is done.

File: src/video/switch/SDL_switchkeyboard.c (hunk 4 above)
Comment (rsn8887) [reply src/video/switch/SDL_switchkeyboard.c]: Done.

File: src/video/switch/SDL_switchmouse.c (hunk 5 above)
Comment (rsn8887) [reply src/video/switch/SDL_switchmouse.c]: Done.

File: src/video/switch/SDL_switchkeyboard.c (hunk 2 above)
Comment (rsn8887) [reply src/video/switch/SDL_switchkeyboard.c]: Done.

File: src/video/switch/SDL_switchkeyboard.c (hunk 1 above)
Comment (rsn8887) [reply src/video/switch/SDL_switchkeyboard.c]: Thanks for the review. This is done.

**Comment by rsn8887** [comment]
I think addressed all the comments. I also added empty SDL_QuitKeyboard() and SDL_QuitMouse() functions, for consistency with other ports.

File: src/video/switch/SDL_switchkeyboard.c (hunk 1 above)
Comment (rsn8887) [reply src/video/switch/SDL_switchkeyboard.c]: Fixed in https://github.com/devkitPro/SDL/pull/33/commits/27115216c99cd97b63af64eda22cb2e5e6034520

File: src/video/switch/SDL_switchkeyboard.c (hunk 2 above)
Comment (rsn8887) [reply src/video/switch/SDL_switchkeyboard.c]: Fixed in https://github.com/devkitPro/SDL/pull/33/commits/27115216c99cd97b63af64eda22cb2e5e6034520 .

File: src/video/switch/SDL_switchkeyboard.c (hunk 3 above)
Comment (rsn8887) [reply src/video/switch/SDL_switchkeyboard.c]: Fixed in https://github.com/devkitPro/SDL/pull/33/commits/27115216c99cd97b63af64eda22cb2e5e6034520

File: src/video/switch/SDL_switchkeyboard.c (hunk 4 above)
Comment (rsn8887) [reply src/video/switch/SDL_switchkeyboard.c]: Fixed in https://github.com/devkitPro/SDL/pull/33/commits/27115216c99cd97b63af64eda22cb2e5e6034520

File: src/video/switch/SDL_switchmouse.c (hunk 5 above)
Comment (rsn8887) [reply src/video/switch/SDL_switchmouse.c]: Fixed in https://github.com/devkitPro/SDL/pull/33/commits/27115216c99cd97b63af64eda22cb2e5e6034520

**Comment by rsn8887** [comment]
Something weird is going on: MousePosition members x and y only report between 0...1280 and 0..720. So either they have to be scaled to 1080p or some other function is missing. velocityX and velocityY seem to depend on the exact time they are being read (they are not integrated).

What is missing is a real measure of "dx" and "dy" to see the difference between last frame and current frame.

Otherwise, mouse motion will not be registered beyond the max and min values of x and y.

**Comment by TurtleP** [comment]
From what I’ve dealt with regarding the mouse coordinates in 1080p, yes they would need to be scaled. However that’s not up to SDL to handle, the end user must scale it as necessary.

**Comment by rsn8887** [comment]
Yes that could be done. 

However, until correct dx and dy values can be read that are not clamped, mouse functionality will be very limited. 

We need a way to get accurate velocities so that mouse motion can be reported even if x and y reach their max or min values.

Current functionality of libnx to read out velocityX and velocityY vie hidMouseRead is broken: It only reports the last velocity values of up to 17 values per frame. See here: 
https://switchbrew.github.io/libnx/structHidMouse.html

We need a nx function to read out all 17 reports to calculate accurate dx and dy values. This would allow relative mouse mode to work.

If that can work, then the clamping could be circumvented, and relative mouse mode would also be possible (currently this is impossible the way libnx is coded).

**Comment by TurtleP** [comment]
However to clarify: if it is the window itself that changed size, SDL would handle that fine. That’s not the case though, as in docked mode it scales the graphics automatically.

**Comment by rsn8887** [comment]
Alternatively hidMouseRead could be modified to report accurate dx, dy values as velocityX and velocityY, by summing over the velocity values of all the new entries.

**Comment by rsn8887** [comment]
Shall we re-scale here or not? It would make life much easier until we have a truly working readout of dx and dy values for mouse and can overcome the clamping to 720p.

**Comment by Cpasjuste** [comment]
I need to take a closer look as my brain is very slow, but i think that if the sdl window reports 1080p (sdl_createwindow (1920..)) and libnx 720p for the mouse then scaling should be handled in sdl2.
For the libnx side i guess some clarification  from @fincs would be usefull.

**Comment by rsn8887** [comment]
I hate configure and it's stupid in-source building.

**Comment by WinterMute** [comment]
@Cpasjuste @rsn8887 how is this kind of thing usually handled in desktop SDL? Should the mouse co-ordinates be dependent on a state set by SDL_RenderSetLogicalSize rather than simple docked 1080p vs undocked 720p?

**Comment by Cpasjuste** [comment]
I really don't  know! We need to look at how other ports handle this (but im away currently :) )

**Comment by rsn8887** [comment]
Hmm the official SDL docs say x and y should be relative to the window: 
https://wiki.libsdl.org/SDL_MouseMotionEvent

So that suggests they should be scaled. This is indeed done in the WinRT implementation. However, what about RendererSetLogicalSize? I think that should also be taken into account. But it seems it is not accounted for at least on WinRT.

**Comment by Cpasjuste** [comment]
I guess it should be handled by the user when RendererSetLogicalSize is used (as the window size doesn't change).

**Comment by rsn8887** [comment]
I found a solution for the clamping issue. The solution is to use velocityX and velocityY and limit readout to not more than once per frame.

fixed in 
https://github.com/devkitPro/SDL/pull/33/commits/2b8f718414f6bf2f67adb78250c4f529ebb884e8

**Comment by rsn8887** [comment]
I am pretty sure this is the correct way to do it, since it also makes relative mouse mode work correctly (for example for first person shooters, where you don't want mouse motion to be limited in any direction)

**Comment by rsn8887** [comment]
This is ready to merge.

**Comment by rsn8887** [comment]
@fincs : Do you want to take another look? Also, shall I squash into a single "switch: add support for physical keyboard and mouse" commit?

**Comment by fincs** [comment]
I'll review it later. Also, we can do the squashing ourselves with the GitHub merge button, don't worry about that.

File: src/video/switch/SDL_switchkeyboard.c [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+#include "SDL_events.h"
+#include "SDL_log.h"
+#include "SDL_switchvideo.h"
+#include "SDL_switchkeyboard.h"
+#include "../../events/SDL_keyboard_c.h"
+
+#define NUM_SCANCODES_SWITCH 160
+
+uint8_t locks = 0;
```
Comment (fincs) [inline src/video/switch/SDL_switchkeyboard.c]: This should be `static`.

File: src/video/switch/SDL_switchkeyboard.c [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+bool keystate[NUM_SCANCODES_SWITCH] = { 0 };
```
Comment (fincs) [inline src/video/switch/SDL_switchkeyboard.c]: This should be `static`.

File: src/video/switch/SDL_switchkeyboard.c [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+uint8_t switch_scancodes[NUM_SCANCODES_SWITCH] = {
```
Comment (fincs) [inline src/video/switch/SDL_switchkeyboard.c]: This should be `static const`.

**Review by fincs — CHANGES_REQUESTED** [review]
.

File: src/video/switch/SDL_switchkeyboard.c (hunk 6 above)
Comment (rsn8887) [reply src/video/switch/SDL_switchkeyboard.c]: fixed in 6201ade922ae72e68aa0e5025c1e91368ab78aed

File: src/video/switch/SDL_switchkeyboard.c (hunk 7 above)
Comment (rsn8887) [reply src/video/switch/SDL_switchkeyboard.c]: fixed in 6201ade922ae72e68aa0e5025c1e91368ab78aed

File: src/video/switch/SDL_switchkeyboard.c (hunk 8 above)
Comment (rsn8887) [reply src/video/switch/SDL_switchkeyboard.c]: fixed in 6201ade922ae72e68aa0e5025c1e91368ab78aed

**Comment by rsn8887** [comment]
Thanks for the review. I made the changes. I think it should be ready now...

File: src/video/switch/SDL_switchkeyboard.c [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+#include "SDL_log.h"
+#include "SDL_switchvideo.h"
+#include "SDL_switchkeyboard.h"
+#include "../../events/SDL_keyboard_c.h"
+
+#define NUM_SCANCODES_SWITCH 160
+
+static uint8_t locks = 0;
+static bool keystate[NUM_SCANCODES_SWITCH] = { 0 };
+static uint8_t switch_scancodes[NUM_SCANCODES_SWITCH] = {
```
Comment (fincs) [inline src/video/switch/SDL_switchkeyboard.c]: This is still non-const. I don't see anywhere where you write to it, so it should be const.

**Review by fincs — CHANGES_REQUESTED** [review]
.

File: src/video/switch/SDL_switchkeyboard.c (hunk 9 above)
Comment (rsn8887) [reply src/video/switch/SDL_switchkeyboard.c]: oops. Fixed in 77066bbdf51bd9f20079612c660e615f648c57cd

**Comment by rsn8887** [comment]
Ok, but now it should finally be ready :)

**Review by fincs — APPROVED** [review]

**Comment by rsn8887** [comment]
@Cpasjuste You can squash and merge this now, whenever you have time.

**Pull request merged**